src/java.base/share/classes/java/lang/ClassValue.java

```diff
@@ -476,6 +476,9 @@ synchronized <T> Object readAccess(ClassValue<T> classValue) {
                 if (updated != entry) {
                     put(classValue.identity, updated);
                 }
+                // Add to the cache, to enable the fast path, next time.
```

Comment:
A question: do we only care about `if (updated != entry)` path to add to cache then? Should this code be moved into that block?

Reply:
I think whenever we hit `readAccess`, it means we are missing this item in the cache and are already in this slow path. I pondered this may happen due to cache being too full too, in which case we have `updated == entry` in the backing map, so leaving it out of the if block should be correct.

Reply:
Also, this cache mechanism was authored before JDK 8 introduced fast table cache in `ConcurrentHashMap`. In the long run, we may switch the ClassValueMap to a ConcurrentHashMap-derived structure, like a `ReferencedKeyMap`.

Reply:
This was the code on JDK 24 GA:
https://github.com/openjdk/jdk/blob/6705a9255d28f351950e7fbca9d05e73942a4e27/src/java.base/share/classes/java/lang/ClassValue.java#L458-L468

This fix is essentially the same as the previous. The cache entry + version is like a bloom filter: if it exists and the version is right, we can ensure we get the right item. A cache entry with an invalid version is essentially invalid, but it stays in the cache unless accessed by cache update requests. Yes I know this may constitute a resource leak, which is tracked by https://bugs.openjdk.org/browse/JDK-8352622.

Reply:
OK, thanks, that is convincing.